src/app/lecture/knowledge-node/notes/markdown-editor/markdown-editor.component.ts

```diff
@@ -0,0 +1,137 @@
+import { Component, Input, ViewChild, forwardRef } from '@angular/core';
+import { ControlValueAccessor, NG_VALUE_ACCESSOR } from '@angular/forms';
+
+@Component({
+  selector: 'cc-markdown-editor',
+  templateUrl: './markdown-editor.component.html',
+  styleUrls: ['./markdown-editor.component.css'],
+  providers: [
+    {
+      provide: NG_VALUE_ACCESSOR,
+      useExisting: forwardRef(() => MarkdownEditorComponent),
+      multi: true
+    }
+  ]
+})
+export class MarkdownEditorComponent implements ControlValueAccessor {
+
+  text = '';
+  livePreview = true;
+  selection: any;
+  @ViewChild('textAreaElement') textArea: HTMLTextAreaElement;
+  @Input() mode = 'edit';
+
+  constructor() { }
+
+  get value(): string {
+    return this.text;
+  }
+
+  set value(value: string) {
+    this.text = value;
+    this.updateChanges();
+  }
+
+  updateChanges(): void {
+    this.onChange(this.value);
+    this.onTouched();
+  }
+
+  onChange: any = () => {};
+
+  onTouched: any = () => {};
+
+  writeValue(obj: any): void {
+    if (obj) {
+      this.value = obj;
+    } else {
+      this.value = '';
+    }
+  }
+
+  registerOnChange(fn: any): void {
+    this.onChange = fn;
+  }
+
+  registerOnTouched(fn: any): void {
+    this.onTouched = fn;
+  }
+
+  setDisabledState?(isDisabled: boolean): void {
+    this.textArea.disabled = isDisabled;
+  }
+
+  insertElement(type: string): void {
+    let tagBegin = '';
+    let tagEnd = '';
+    let tagText = '';
```

Comment:
A minor thing to add if it is easy - once a button is clicked and the text is added, can we select the text? E.g., when I click bold it writes **Bold Text** and has Bold Text selected so that I can instantly write over it?

Of the top of my head, you would need to find a way to focus the text area ([this](https://stackoverflow.com/questions/54557400/how-to-focus-a-textarea-that-has-an-ngif) might help) and select its content from index tagBegin length to index content.length - tagEnd.length.